File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/sso/KeycloakAPIClient.java
package com.wearezeta.auto.common.sso;

import com.wearezeta.auto.common.backend.BackendConnections;
import com.wearezeta.auto.common.backend.HttpRequestException;
import com.wearezeta.auto.common.credentials.Credentials;
import com.wearezeta.auto.common.log.ZetaLogger;
import org.glassfish.jersey.client.ClientConfig;
import org.glassfish.jersey.client.ClientProperties;
import org.json.JSONArray;
import org.json.JSONObject;

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import javax.ws.rs.core.MediaType;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;

import static java.net.HttpURLConnection.*;

public class KeycloakAPIClient {

    private static final Logger log = ZetaLogger.getLog(KeycloakAPIClient.class.getSimpleName());
    private static final Duration CONNECT_TIMEOUT = Duration.ofSeconds(3);
    private static final Duration READ_TIMEOUT = Duration.ofSeconds(240);
    private static final String REALM = "master";
    private static final String ADMIN = "admin";
    private final String backendName;

    private String clientId;
    private Set<String> userIds = new HashSet<>();

    // Create a trust manager that does not validate certificate chains
    TrustManager[] trustAllCerts = new TrustManager[]{
            new X509TrustManager() {
                public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                    return null;
                }

                public void checkClientTrusted(
                        java.security.cert.X509Certificate[] certs, String authType) {
                }

                public void checkServerTrusted(
                        java.security.cert.X509Certificate[] certs, String authType) {
                }
            }
    };

    public KeycloakAPIClient(String backendName) {
        this.backendName = backendName;
        final ClientConfig configuration = new ClientConfig();
        configuration.property(ClientProperties.CONNECT_TIMEOUT, (int) CONNECT_TIMEOUT.toMillis());
        configuration.property(ClientProperties.READ_TIMEOUT, (int) READ_TIMEOUT.toMillis());
        // Install the all-trusting trust manager
        try {
            SSLContext sc = SSLContext.getInstance("SSL");
            sc.init(null, trustAllCerts, new java.security.SecureRandom());
            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
        } catch (Exception e) {
            log.severe("Could not install all-trusting trust manager: " + e.getMessage());
        }
    }

    private String getBaseURI() {
        return BackendConnections.get(backendName).getKeycloakUrl();
    }

    public String getMetadata() {
        HttpURLConnection c = buildDefaultRequest(
                String.format("/realms/%s/protocol/saml/descriptor", REALM),
                MediaType.APPLICATION_XML);
        final String output = httpGet(c, new int[]{HTTP_OK});
        DomUtils.toDocument(output);
        return output;
    }

    public void createSAMLClient(String teamId, String backendURL) {
        String token = authorize();
        log.info("Token: " + token);
        HttpURLConnection c = buildAuthorizedRequest(String.format("admin/realms/%s/clients", REALM),
                MediaType.APPLICATION_JSON, token);
        String finalizeUrl = String.format("%ssso/finalize-login/%s", backendURL, teamId);
        JSONObject requestBody = new JSONObject();
        requestBody.put("clientId", finalizeUrl);
        //requestBody.put("surrogateAuthRequired", false);
        requestBody.put("enabled", true);
        requestBody.put("adminUrl", "");
        requestBody.put("baseUrl", "");
        requestBody.put("rootUrl", "");
        requestBody.put("name", "");
        requestBody.put("description", "");
        //requestBody.put("clientAuthenticatorType", "client-secret");
        //requestBody.put("secret", "iJg6Ysa0qWdmGCnNzLkqb6rTjjhRgPF6");
        JSONArray redirectUris = new JSONArray();
        redirectUris.put(finalizeUrl);
        requestBody.put("redirectUris", redirectUris);
        JSONArray webOrigins = new JSONArray();
        webOrigins.put(backendURL.substring(0, backendURL.length() - 1));
        requestBody.put("webOrigins", webOrigins);
        requestBody.put("protocol", "saml");
        JSONObject attributes = new JSONObject();
        attributes.put("display.on.consent.screen", "false");
        attributes.put("saml.encrypt", "false");
        attributes.put("saml_assertion_consumer_url_post", finalizeUrl);
        attributes.put("saml.client.signature", "false");
        attributes.put("saml.artifact.binding", "false");
        attributes.put("saml.assertion.signature", "true");
        attributes.put("saml.onetimeuse.condition", "false");
        attributes.put("saml.server.signature.keyinfo.ext", "false");
        attributes.put("saml.server.signature.keyinfo.xmlSigKeyInfoKeyNameTransformer", "NONE");
        requestBody.put("attributes", attributes);
        String location = httpPost(c, requestBody.toString(), new int[]{HTTP_CREATED});
        this.clientId = getIdFromLocation(location);
    }

    public void createUser(String username, String firstname, String lastname, String email, String password) {
        HttpURLConnection c = buildAuthorizedRequest(String.format("admin/realms/%s/users", REALM),
                MediaType.APPLICATION_JSON, authorize());
        JSONObject user = new JSONObject();
        user.put("username", username);
        user.put("firstName", firstname);
        user.put("lastName", lastname);
        user.put("email", email);
        user.put("emailVerified", true);
        user.put("enabled", true);
        JSONArray credentials = new JSONArray();
        JSONObject passwordCredential = new JSONObject();
        passwordCredential.put("type", "password");
        passwordCredential.put("value", password);
        passwordCredential.put("temporary", false);
        credentials.put(passwordCredential);
        user.put("credentials", credentials);
        String location = httpPost(c, user.toString(), new int[]{HTTP_CREATED});
        userIds.add(getIdFromLocation(location));
    }

    public void cleanUp() {
        if (clientId != null) {
            deleteSAMLClient(clientId);
        }
        for (String userId : userIds) {
            deleteUser(userId);
        }
    }

    private void deleteSAMLClient(String clientId) {
        HttpURLConnection c = buildAuthorizedRequest(String.format("admin/realms/%s/clients/%s", REALM, clientId),
                MediaType.APPLICATION_JSON, authorize());
        httpDelete(c, new int[]{HTTP_NO_CONTENT});
    }

    private void deleteUser(String userId) {
        HttpURLConnection c = buildAuthorizedRequest(String.format("admin/realms/%s/users/%s", REALM, userId),
                MediaType.APPLICATION_JSON, authorize());
        httpDelete(c, new int[]{HTTP_NO_CONTENT});
    }

    private String getIdFromLocation(String location) {
        log.info("Location: " + location);
        return location.substring(location.lastIndexOf("/") + 1);
    }

    private String authorize() {
        String PASSWORD = Credentials.get("KEYCLOAK_PASSWORD");
        HttpURLConnection c = null;
        try {
            URL url = new URL(String.format("%s/realms/%s/protocol/openid-connect/token", getBaseURI(), REALM));
            c = (HttpURLConnection) url.openConnection();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        c.setRequestProperty("Content-Type", MediaType.APPLICATION_FORM_URLENCODED);
        c.setRequestProperty("Accept", MediaType.APPLICATION_JSON);
        String response = "";
        int status = -1;
        int[] acceptableResponseCodes = new int[]{HTTP_OK};
        try {
            String requestBody = "client_id=" + URLEncoder.encode("admin-cli", StandardCharsets.UTF_8) +
                    "&username=" + URLEncoder.encode(ADMIN, StandardCharsets.UTF_8) +
                    "&password=" + URLEncoder.encode(PASSWORD, StandardCharsets.UTF_8) +
                    "&grant_type=" + URLEncoder.encode("password", StandardCharsets.UTF_8);
            log.info("POST " + c.getURL());
            c.setRequestMethod("POST");
            logHttpRequestProperties(c);
            logRequest(requestBody);
            c.setDoOutput(true);
            writeStream(requestBody, c.getOutputStream());
            status = c.getResponseCode();
            response = readStream(c.getInputStream());
            logResponseAndStatusCode(response, status);
            assertResponseCode(status, acceptableResponseCodes);
            return new JSONObject(response).getString("access_token");
        } catch (IOException e) {
            try {
                response = readStream(c.getErrorStream());
            } catch (IOException ex) {
                log.fine("Could not read error stream: " + e.getMessage());
            }
            if (Arrays.stream(acceptableResponseCodes).anyMatch(acceptable -> acceptable > 400)) {
                assertResponseCode(status, acceptableResponseCodes);
                log.info(String.format(">>> Response (%s): %s", status, response));
                return response;
            } else {
                String error = String.format("%s (%s): %s", e.getMessage(), status, response);
                log.severe(error);
                throw new HttpRequestException(error, status);
            }
        } finally {
            c.disconnect();
        }
    }

    private HttpURLConnection buildAuthorizedRequest(String path, String mediaType, String token) {
        HttpURLConnection c = null;
        try {
            URL url = new URL(String.format("%s/%s", getBaseURI(), path));
            c = (HttpURLConnection) url.openConnection();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        c.setRequestProperty("Content-Type", mediaType);
        c.setRequestProperty("Accept", mediaType);
        c.setRequestProperty("Authorization", "Bearer " + token);
        return c;
    }

    private HttpURLConnection buildDefaultRequest(String path, String mediaType) {
        HttpURLConnection c = null;
        try {
            URL url = new URL(String.format("%s%s", getBaseURI(), path));
            c = (HttpURLConnection) url.openConnection();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        c.setRequestProperty("Content-Type", mediaType);
        c.setRequestProperty("Accept", mediaType);
        return c;
    }

    private String httpGet(HttpURLConnection c, int[] acceptableResponseCodes) {
        String response = "";
        int status = -1;
        try {
            log.info("GET " + c.getURL());
            c.setRequestMethod("GET");
            logHttpRequestProperties(c);
            status = c.getResponseCode();
            response = readStream(c.getInputStream());
            logResponseAndStatusCode(response, status);
            assertResponseCode(status, acceptableResponseCodes);
            return response;
        } catch (IOException e) {
            try {
                response = readStream(c.getErrorStream());
            } catch (IOException ex) {
                log.fine("Could not read error stream: " + e.getMessage());
            }
            if (Arrays.stream(acceptableResponseCodes).anyMatch(acceptable -> acceptable > 400)) {
                assertResponseCode(status, acceptableResponseCodes);
                log.info(String.format(">>> Response (%s): %s", status, response));
                return response;
            } else {
                String error = String.format("%s (%s): %s", e.getMessage(), status, response);
                log.severe(error);
                throw new HttpRequestException(error, status);
            }
        } finally {
            c.disconnect();
        }
    }

    private String httpPost(HttpURLConnection c, String requestBody, int[] acceptableResponseCodes) {
        String response = "";
        String location = "";
        int status = -1;
        try {
            log.info("POST " + c.getURL());
            c.setRequestMethod("POST");
            logHttpRequestProperties(c);
            logRequest(requestBody);
            c.setDoOutput(true);
            writeStream(requestBody, c.getOutputStream());
            status = c.getResponseCode();
            response = readStream(c.getInputStream());
            location = c.getHeaderField("Location");
            logResponseAndStatusCode(response, status);
            assertResponseCode(status, acceptableResponseCodes);
            return location;
        } catch (IOException e) {
            try {
                response = readStream(c.getErrorStream());
            } catch (IOException ex) {
                log.fine("Could not read error stream: " + e.getMessage());
            }
            if (Arrays.stream(acceptableResponseCodes).anyMatch(acceptable -> acceptable > 400)) {
                assertResponseCode(status, acceptableResponseCodes);
                log.info(String.format(">>> Response (%s): %s", status, response));
                return response;
            } else {
                String error = String.format("%s (%s): %s", e.getMessage(), status, response);
                log.severe(error);
                throw new HttpRequestException(error, status);
            }
        } finally {
            c.disconnect();
        }
    }

    private String httpDelete(HttpURLConnection c, int[] acceptableResponseCodes) {
        String response = "";
        int status = -1;
        try {
            log.info("DELETE " + c.getURL());
            c.setRequestMethod("DELETE");
            logHttpRequestProperties(c);
            c.setDoOutput(true);
            writeStream("", c.getOutputStream());
            status = c.getResponseCode();
            response = readStream(c.getInputStream());
            logResponseAndStatusCode(response, status);
            assertResponseCode(status, acceptableResponseCodes);
            return response;
        } catch (IOException e) {
            try {
                response = readStream(c.getErrorStream());
            } catch (IOException ex) {
                log.fine("Could not read error stream: " + e.getMessage());
            }
            String error = String.format("%s (%s): %s", e.getMessage(), status, response);
            log.severe(error);
            throw new HttpRequestException(error, status);
        } finally {
            c.disconnect();
        }
    }

    private void writeStream(String data, OutputStream os) throws IOException {
        DataOutputStream wr = new DataOutputStream(os);
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(wr, StandardCharsets.UTF_8));
        try {
            writer.write(data);
        } finally {
            writer.close();
            wr.close();
        }
    }

    private String readStream(InputStream is) throws IOException {
        if (is != null) {
            try (BufferedReader in = new BufferedReader(new InputStreamReader(is))) {
                String inputLine;
                StringBuilder content = new StringBuilder();
                while ((inputLine = in.readLine()) != null) {
                    content.append(inputLine);
                }
                return content.toString();
            }
        }
        return "";
    }

    private void logRequest(String request) {
        if (request.isEmpty()) {
            log.info(" >>> Request with no request body");
        } else {
            if (log.isLoggable(Level.FINE)) {
                log.info(String.format(" >>> Request: %s", request));
            } else {
                log.info(String.format(" >>> Request: %s", truncate(request)));
            }
        }
    }

    private void logResponseAndStatusCode(String response, int responseCode) {
        if (response.isEmpty()) {
            log.info(String.format(" >>> Response (%s) with no response body", responseCode));
        } else {
            if (log.isLoggable(Level.FINE)) {
                log.info(String.format(" >>> Response (%s): %s", responseCode, response));
            } else {
                log.info(String.format(" >>> Response (%s): %s", responseCode, truncate(response)));
            }
        }
    }

    private String truncate(String text) {
        final int MAX_LOG_ENTRY_LENGTH = 280;
        if (text.length() > MAX_LOG_ENTRY_LENGTH) {
            return text.substring(0, MAX_LOG_ENTRY_LENGTH) + "...";
        }
        return text;
    }

    private void assertResponseCode(int responseCode, int[] acceptableResponseCodes) {
        if (Arrays.stream(acceptableResponseCodes).noneMatch(a -> a == responseCode)) {
            throw new HttpRequestException(
                    String.format("Backend request failed. Request return code is: %d. Expected codes are: %s.",
                            responseCode,
                            Arrays.toString(acceptableResponseCodes)),
                    responseCode);
        }
    }

    private void logHttpRequestProperties(HttpURLConnection c) {
        if (log.isLoggable(Level.FINE)) {
            for (String property : c.getRequestProperties().keySet()) {
                List<String> values = Collections.singletonList(c.getRequestProperty(property));
                log.fine(String.format("%s: %s", property, String.join(", ", values)));
            }
        }
    }
}

File: tests/testsCore/src/main/kotlin/com/wire/android/tests/core/config/KeycloakAPIClient.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
package com.wire.android.tests.core.config


import com.wire.android.tests.core.exceptions.HttpRequestException
import com.wire.android.tests.core.services.backend.BackendConnections
import com.wire.android.tests.core.utils.DomUtils
import com.wire.android.tests.core.utils.ZetaLogger
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.glassfish.jersey.client.ClientConfig
import org.glassfish.jersey.client.ClientProperties
import org.json.JSONArray
import org.json.JSONObject
import java.io.*
import java.net.HttpURLConnection
import java.net.HttpURLConnection.*
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.time.Duration
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager
import javax.ws.rs.core.MediaType


class KeycloakAPIClient {

    val log: Logger = ZetaLogger.getLog(KeycloakAPIClient::class.simpleName)
    val CONNECT_TIMEOUT: Duration = Duration.ofSeconds(3)
    val READ_TIMEOUT: Duration = Duration.ofSeconds(240)
    val REALM: String = "master"
    val ADMIN: String = "admin"
    private var backendName: String? = null

    private var clientId: String? = null
    private var userIds = mutableSetOf<String>()

    // Create a trust manager that does not validate certificate chains
    val trustAllCerts: Array<TrustManager> = arrayOf(object : X509TrustManager {
        override fun getAcceptedIssuers(): Array<X509Certificate> {
            return emptyArray()  // Return empty array instead of null
        }

        override fun checkClientTrusted(
            certs: Array<X509Certificate>,
            authType: String
        ) {
            // Trust all client certificates
        }

        override fun checkServerTrusted(
            certs: Array<X509Certificate>,
            authType: String
        ) {
            // Trust all server certificates
        }
    })

    fun KeycloakAPIClient(backendName: String?) {
        this.backendName = backendName
        val configuration = ClientConfig()
        configuration.property(ClientProperties.CONNECT_TIMEOUT, CONNECT_TIMEOUT.toMillis().toInt())
        configuration.property(ClientProperties.READ_TIMEOUT, READ_TIMEOUT.toMillis().toInt())
        // Install the all-trusting trust manager
        try {
            val sc = SSLContext.getInstance("SSL")
            sc.init(null, trustAllCerts, SecureRandom())
            HttpsURLConnection.setDefaultSSLSocketFactory(sc.socketFactory)
        } catch (e: Exception) {
            log.severe("Could not install all-trusting trust manager: " + e.message)
        }
    }

    private fun getBaseURI(): String {
        val connections = BackendConnections()
        return connections.get(backendName).getKeycloakUrl()
    }

    fun getMetadata(): String {
        val c: HttpURLConnection? = buildDefaultRequest(
            String.format("/realms/%s/protocol/saml/descriptor", REALM),
            MediaType.APPLICATION_XML
        )
        if (c == null){
            throw HttpRequestException("url connection is null")
        }
        val output: String = httpGet(c, intArrayOf(HTTP_OK))
        val domUtils =DomUtils()
        domUtils.toDocument(output)
        return output
    }

    fun createSAMLClient(teamId: String?, backendURL: String) {
        val token: String = authorize()
        log.info("Token: $token")
        val c: HttpURLConnection? = buildAuthorizedRequest(
            String.format("admin/realms/%s/clients", REALM),
            MediaType.APPLICATION_JSON, token
        )
        val finalizeUrl = String.format("%ssso/finalize-login/%s", backendURL, teamId)
        val requestBody = JSONObject()
        requestBody.put("clientId", finalizeUrl)
        //requestBody.put("surrogateAuthRequired", false);
        requestBody.put("enabled", true)
        requestBody.put("adminUrl", "")
        requestBody.put("baseUrl", "")
        requestBody.put("rootUrl", "")
        requestBody.put("name", "")
        requestBody.put("description", "")
        //requestBody.put("clientAuthenticatorType", "client-secret");
        //requestBody.put("secret", "iJg6Ysa0qWdmGCnNzLkqb6rTjjhRgPF6");
        val redirectUris = JSONArray()
        redirectUris.put(finalizeUrl)
        requestBody.put("redirectUris", redirectUris)
        val webOrigins = JSONArray()
        webOrigins.put(backendURL.substring(0, backendURL.length - 1))
        requestBody.put("webOrigins", webOrigins)
        requestBody.put("protocol", "saml")
        val attributes = JSONObject()
        attributes.put("display.on.consent.screen", "false")
        attributes.put("saml.encrypt", "false")
        attributes.put("saml_assertion_consumer_url_post", finalizeUrl)
        attributes.put("saml.client.signature", "false")
        attributes.put("saml.artifact.binding", "false")
        attributes.put("saml.assertion.signature", "true")
        attributes.put("saml.onetimeuse.condition", "false")
        attributes.put("saml.server.signature.keyinfo.ext", "false")
        attributes.put("saml.server.signature.keyinfo.xmlSigKeyInfoKeyNameTransformer", "NONE")
        requestBody.put("attributes", attributes)
        if (c == null){
            throw HttpRequestException("url connection is null")
        }
        val location: String = httpPost(c, requestBody.toString(), intArrayOf(HTTP_CREATED))
        this.clientId = getIdFromLocation(location)
    }

    fun createSAMLClientQARealm(teamId: String?, backendURL: String) {
        val REALM = "QA"
        val token: String = authorize()
        log.info("Token: $token")
        val c: HttpURLConnection? = buildAuthorizedRequest(
            String.format("admin/realms/%s/clients", REALM),
            MediaType.APPLICATION_JSON, token
        )
        val finalizeUrl = String.format("%ssso/finalize-login/%s", backendURL, teamId)
        val requestBody = JSONObject()
        requestBody.put("clientId", finalizeUrl)
        //requestBody.put("surrogateAuthRequired", false);
        requestBody.put("enabled", true)
        requestBody.put("adminUrl", "")
        requestBody.put("baseUrl", "")
        requestBody.put("rootUrl", "")
        requestBody.put("name", "")
        requestBody.put("description", "")
        //requestBody.put("clientAuthenticatorType", "client-secret");
        //requestBody.put("secret", "iJg6Ysa0qWdmGCnNzLkqb6rTjjhRgPF6");
        val redirectUris = JSONArray()
        redirectUris.put(finalizeUrl)
        requestBody.put("redirectUris", redirectUris)
        val webOrigins = JSONArray()
        webOrigins.put(backendURL.substring(0, backendURL.length - 1))
        requestBody.put("webOrigins", webOrigins)
        requestBody.put("protocol", "saml")
        val attributes = JSONObject()
        attributes.put("display.on.consent.screen", "false")
        attributes.put("saml.encrypt", "false")
        attributes.put("saml_assertion_consumer_url_post", finalizeUrl)
        attributes.put("saml.client.signature", "false")
        attributes.put("saml.artifact.binding", "false")
        attributes.put("saml.assertion.signature", "true")
        attributes.put("saml.onetimeuse.condition", "false")
        attributes.put("saml.server.signature.keyinfo.ext", "false")
        attributes.put("saml.server.signature.keyinfo.xmlSigKeyInfoKeyNameTransformer", "NONE")
        requestBody.put("attributes", attributes)
        if (c == null){
            throw HttpRequestException("url connection is null")
        }
        val location: String = httpPost(c, requestBody.toString(), intArrayOf(HTTP_CREATED))
        this.clientId = getIdFromLocation(location)
    }

    fun createUser(username: String?, firstname: String?, lastname: String?, email: String?, password: String?) {
        val c: HttpURLConnection? = buildAuthorizedRequest(
            String.format("admin/realms/%s/users", REALM),
            MediaType.APPLICATION_JSON, authorize()
        )
        val user = JSONObject()
        user.put("username", username)
        user.put("firstName", firstname)
        user.put("lastName", lastname)
        user.put("email", email)
        user.put("emailVerified", true)
        user.put("enabled", true)
        val credentials = JSONArray()
        val passwordCredential = JSONObject()
        passwordCredential.put("type", "password")
        passwordCredential.put("value", password)
        passwordCredential.put("temporary", false)
        credentials.put(passwordCredential)
        user.put("credentials", credentials)
        if (c == null){
            throw HttpRequestException("url connection is null")
        }
        val location: String = httpPost(c, user.toString(), intArrayOf(HTTP_CREATED))
        userIds.add(getIdFromLocation(location))
    }

    fun createUserQARealm(username: String?, firstname: String?, lastname: String?, email: String?, password: String?) {
        val REALM = "QA"
        val c: HttpURLConnection? = buildAuthorizedRequest(
            String.format("admin/realms/%s/users", REALM),
            MediaType.APPLICATION_JSON, authorize()
        )
        val user = JSONObject()
        user.put("username", username)
        user.put("firstName", firstname)
        user.put("lastName", lastname)
        user.put("email", email)
        user.put("emailVerified", true)
        user.put("enabled", true)
        val credentials = JSONArray()
        val passwordCredential = JSONObject()
        passwordCredential.put("type", "password")
        passwordCredential.put("value", password)
        passwordCredential.put("temporary", false)
        credentials.put(passwordCredential)
        user.put("credentials", credentials)
        if (c == null){
            throw HttpRequestException("url connection is null")
        }
        val location: String = httpPost(c, user.toString(), intArrayOf(HTTP_CREATED))
        userIds.add(getIdFromLocation(location))
    }

    fun cleanUp() {
        clientId?.let { deleteSAMLClient(it) }
        for (userId in userIds) {
            deleteUser(userId)
        }
    }

    fun cleanUpQARealm() {
        val REALM = "QA"
        val token: String = authorize()
        //        JSONObject requestBody = new JSONObject();
//        JSONObject user = new JSONObject();
        val client: HttpURLConnection? =
            buildAuthorizedRequest(String.format("admin/realms/%s/clients", REALM), MediaType.APPLICATION_JSON, token)
        val userConnection: HttpURLConnection? =
            buildAuthorizedRequest(String.format("admin/realms/%s/users", REALM), MediaType.APPLICATION_JSON, token)
        if (client == null){
            throw HttpRequestException("url connection is null")
        }
        val location: String = httpGet(client, intArrayOf(HTTP_CREATED))
        clientId = getIdFromLocation(location)
        if (userConnection == null){
            throw HttpRequestException("url connection is null")
        }
        val locationUser: String = httpGet(userConnection, intArrayOf(HTTP_CREATED))
        userIds = mutableSetOf(getIdFromLocation(locationUser))
        clientId?.let { deleteSAMLClientQA(it) }
        for (userId in userIds) {
            deleteUserQA(userId)
        }
    }


    private fun deleteSAMLClient(clientId: String) {
        val c = buildAuthorizedRequest(
            String.format("admin/realms/%s/clients/%s", REALM, clientId),
            MediaType.APPLICATION_JSON, authorize()
        )
        httpDelete(c, intArrayOf(HTTP_NO_CONTENT))
    }

    private fun deleteSAMLClientQA(clientId: String) {
        val REALM = "QA"
        val c = buildAuthorizedRequest(
            String.format("admin/realms/%s/clients/%s", REALM, clientId),
            MediaType.APPLICATION_JSON, authorize()
        )
        httpDelete(c, intArrayOf(HTTP_NO_CONTENT))
    }

    private fun deleteUser(userId: String) {
        val c = buildAuthorizedRequest(
            String.format("admin/realms/%s/users/%s", REALM, userId),
            MediaType.APPLICATION_JSON, authorize()
        )
        httpDelete(c, intArrayOf(HTTP_NO_CONTENT))
    }

    private fun deleteUserQA(userId: String) {
        val REALM = "QA"
        val c = buildAuthorizedRequest(
            String.format("admin/realms/%s/users/%s", REALM, userId),
            MediaType.APPLICATION_JSON, authorize()
        )
        httpDelete(c, intArrayOf(HTTP_NO_CONTENT))
    }

    private fun getIdFromLocation(location: String): String {
        log.info("Location: $location")
        return location.substring(location.lastIndexOf("/") + 1)
    }

    private fun authorize(): String {
        val credentials = Credentials()
        val PASSWORD: String = credentials.getCredentials("KEYCLOAK_PASSWORD")
        val client = OkHttpClient()

        val requestBody = FormBody.Builder()
            .add("client_id", "admin-cli")
            .add("username", ADMIN)
            .add("password", PASSWORD)
            .add("grant_type", "password")
            .build()

        val request = Request.Builder()
            .url("${getBaseURI()}/realms/master/protocol/openid-connect/token")
            .post(requestBody)
            .addHeader("Content-Type", MediaType.APPLICATION_FORM_URLENCODED)
            .addHeader("Accept", MediaType.APPLICATION_JSON)
            .build()

        log.info("POST ${request.url}")
        logHttpRequestHeaders(request)

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string() ?: ""
            val status = response.code
            logResponseAndStatusCode(responseBody, status)

            if (!response.isSuccessful) {
                val error = "${response.message} ($status): $responseBody"
                log.severe(error)
                throw HttpRequestException(error, status)
            }

            return JSONObject(responseBody).getString("access_token")
        }
    }

    // Helper function to log request headers (similar to logHttpRequestProperties)
    private fun logHttpRequestHeaders(request: Request) {
        val headers = request.headers
        for (i in 0 until headers.size) {
            log.info(">>> ${headers.name(i)}: ${headers.value(i)}")
        }
    }

    private fun buildAuthorizedRequest(path: String, mediaType: String, token: String): HttpURLConnection? {
        var c: HttpURLConnection? = null
        try {
            val url = URL(String.format("%s/%s", getBaseURI(), path))
            c = url.openConnection() as HttpURLConnection
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        c!!.setRequestProperty("Content-Type", mediaType)
        c!!.setRequestProperty("Accept", mediaType)
        c!!.setRequestProperty("Authorization", "Bearer $token")
        return c
    }

    private fun buildDefaultRequest(path: String, mediaType: String): HttpURLConnection? {
        var c: HttpURLConnection? = null
        try {
            val url = URL(String.format("%s%s", getBaseURI(), path))
            c = url.openConnection() as HttpURLConnection
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        c!!.setRequestProperty("Content-Type", mediaType)
        c!!.setRequestProperty("Accept", mediaType)
        return c
    }

    private fun httpGet(c: HttpURLConnection, acceptableResponseCodes: IntArray): String {
        var response = ""
        var status = -1
        try {
            log.info("GET " + c.url)
            c.requestMethod = "GET"
            logHttpRequestProperties(c)
            status = c.responseCode
            response = readStream(c.inputStream)
            logResponseAndStatusCode(response, status)
            assertResponseCode(status, acceptableResponseCodes)
            return response
        } catch (e: IOException) {
            try {
                response = readStream(c.errorStream)
            } catch (ex: IOException) {
                log.fine("Could not read error stream: " + e.message)
            }
            if (Arrays.stream(acceptableResponseCodes).anyMatch { acceptable: Int -> acceptable > 400 }) {
                assertResponseCode(status, acceptableResponseCodes)
                log.info(String.format(">>> Response (%s): %s", status, response))
                return response
            } else {
                val error = String.format("%s (%s): %s", e.message, status, response)
                log.severe(error)
                throw HttpRequestException(error, status)
            }
        } finally {
            c.disconnect()
        }
    }

    private fun httpPost(c: HttpURLConnection, requestBody: String, acceptableResponseCodes: IntArray): String {
        var response = ""
        var location = ""
        var status = -1
        try {
            log.info("POST " + c.url)
            c.requestMethod = "POST"
            logHttpRequestProperties(c)
            logRequest(requestBody)
            c.doOutput = true
            writeStream(requestBody, c.outputStream)
            status = c.responseCode
            response = readStream(c.inputStream)
            location = c.getHeaderField("Location")
            logResponseAndStatusCode(response, status)
            assertResponseCode(status, acceptableResponseCodes)
            return location
        } catch (e: IOException) {
            try {
                response = readStream(c.errorStream)
            } catch (ex: IOException) {
                log.fine("Could not read error stream: " + e.message)
            }
            if (Arrays.stream(acceptableResponseCodes).anyMatch { acceptable: Int -> acceptable > 400 }) {
                assertResponseCode(status, acceptableResponseCodes)
                log.info(String.format(">>> Response (%s): %s", status, response))
                return response
            } else {
                val error = String.format("%s (%s): %s", e.message, status, response)
                log.severe(error)
                throw HttpRequestException(error, status)
            }
        } finally {
            c.disconnect()
        }
    }

    private fun httpDelete(c: HttpURLConnection?, acceptableResponseCodes: IntArray): String {
        var response = ""
        var status = -1
        try {
            log.info("DELETE " + c!!.url)
            c!!.requestMethod = "DELETE"
            logHttpRequestProperties(c)
            c!!.doOutput = true
            writeStream("", c!!.outputStream)
            status = c!!.responseCode
            response = readStream(c!!.inputStream)
            logResponseAndStatusCode(response, status)
            assertResponseCode(status, acceptableResponseCodes)
            return response
        } catch (e: IOException) {
            try {
                response = readStream(c!!.errorStream)
            } catch (ex: IOException) {
                log.fine("Could not read error stream: " + e.message)
            }
            val error = String.format("%s (%s): %s", e.message, status, response)
            log.severe(error)
            throw HttpRequestException(error, status)
        } finally {
            c!!.disconnect()
        }
    }

    @Throws(IOException::class)
    private fun writeStream(data: String, os: OutputStream) {
        val wr = DataOutputStream(os)
        val writer = BufferedWriter(OutputStreamWriter(wr, StandardCharsets.UTF_8))
        try {
            writer.write(data)
        } finally {
            writer.close()
            wr.close()
        }
    }

    @Throws(IOException::class)
    private fun readStream(`is`: InputStream?): String {
        if (`is` != null) {
            BufferedReader(InputStreamReader(`is`)).use { `in` ->
                var inputLine: String?
                var content = ""
                while ((`in`.readLine().also { inputLine = it }) != null) {
                    content += inputLine
                }
                return content
            }
        }
        return ""
    }

    private fun logRequest(request: String) {
        if (request.isEmpty()) {
            log.info(" >>> Request with no request body")
        } else {
            if (log.isLoggable(Level.FINE)) {
                log.info(String.format(" >>> Request: %s", request))
            } else {
                log.info(String.format(" >>> Request: %s", truncate(request)))
            }
        }
    }

    private fun logResponseAndStatusCode(response: String, responseCode: Int) {
        if (response.isEmpty()) {
            log.info(String.format(" >>> Response (%s) with no response body", responseCode))
        } else {
            if (log.isLoggable(Level.FINE)) {
                log.info(String.format(" >>> Response (%s): %s", responseCode, response))
            } else {
                log.info(String.format(" >>> Response (%s): %s", responseCode, truncate(response)))
            }
        }
    }

    private fun truncate(text: String): String {
        val MAX_LOG_ENTRY_LENGTH = 280
        if (text.length > MAX_LOG_ENTRY_LENGTH) {
            return text.substring(0, MAX_LOG_ENTRY_LENGTH) + "..."
        }
        return text
    }

    private fun assertResponseCode(responseCode: Int, acceptableResponseCodes: IntArray) {
        if (Arrays.stream(acceptableResponseCodes).noneMatch { a: Int -> a == responseCode }) {
            throw HttpRequestException(
                String.format(
                    "Backend request failed. Request return code is: %d. Expected codes are: %s.",
                    responseCode,
                    acceptableResponseCodes.contentToString()
                ),
                responseCode
            )
        }
    }

    private fun logHttpRequestProperties(c: HttpURLConnection?) {
        if (log.isLoggable(Level.FINE)) {
            for (property in c!!.requestProperties.keys) {
                val values = listOf(c!!.getRequestProperty(property))
                log.fine(String.format("%s: %s", property, java.lang.String.join(", ", values)))
            }
        }
    }
}
